lib/web/desktop.go

```diff
@@ -504,25 +498,27 @@ func proxyWebsocketConn(ctx context.Context, ws *websocket.Conn, wds net.Conn, l
 
 	tdpConnProxy := tdp.NewConnProxy(&WebsocketIO{Conn: ws}, wds, func(_ *tdp.Conn, msg tdp.Message) (tdp.Message, error) {
 		if ping, ok := msg.(tdp.Ping); ok {
-			pings <- ping
+			select {
+			case pings <- ping:
+			case <-ctx.Done():
+			}
 			return nil, nil
 		}
-
-		if ls, ok := msg.(tdp.LatencyStats); ok {
-			log.DebugContext(ctx, "sending latency stats", "client", ls.ClientLatency, "server", ls.ServerLatency)
-		}
 		return msg, nil
 	})
 
 	if latencySupported {
+		conn := tdp.NewConn(wds)
+		defer conn.Close()
```

Comment:
Using two distinct `*tdp.Conn`s bypasses the write lock in `tdp.Conn` so we're still relying on `(*tls.Conn).Write` to not split the buffers given to it (which, while perfectly reasonable and unlikely to ever change, is not documented anywhere) and `(*ConnProxy).Run` actually does a bytestream `io.Copy` from the client to the server so unless i misunderstood something this can definitely shove the encoding of a `tdp.Ping` right in the middle of some other message.

If `*ConnProxy` needs to be able to parse and inject messages I think that both directions must be aware of the framing (not necessarily fully decoding them, mind you, but I don't know enough about the protocol to know if it's possible to "half-parse" it).

Reply:
I do something similar in https://github.com/gravitational/teleport/pull/54795/commits/bafa94dae5474a1c62a77311decae50b6020ae7e now, let me know WDYT.
We don't have half-parse option unfortunately so we have to parse it in full. We might add length calculation to messages so we can copy them without fully parsing but it's out of scope for this PR I think